### PR TITLE
feat: add IntegrationOptionsPlugin

### PR DESCRIPTION
**Pull request by vgupta98**
## Description

In this PR, `IntegrationOptionsPlugin` is added which is responsible for filtering out events for integrations based on user set integration options for events.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Implementation Details

A new plugin called `IntegrationOptionsPlugin` is added. It is added in the `PluginChain` instance of `IntegrationPlugin` abstract class. It will be a default plugin for all the integrations since client side event filtering is a common feature for all the integrations.
The plugin checks if the destination is explicitly enabled or disabled AND if all destinations are enabled or disabled in the event's integrations and then passes or drops the event accordingly. Tests for `IntegrationOptionsPlugin` are also added.
### Logic for Integrations options:
It applies the below logic:
All -> true - Allow
All -> false - Block

All -> false && key = true - Allow
All -> false && key = false - Block

key = true - Allow
key = false - Block
Here, `All` key represents all the destinations and `key` represents the key for a specific destination. The value for `key` is given preference over value for `All`.

Note: The `integrations` is a JsonObject therefore, it can be set to any complex JsonObject. The logic for IntegrationOptionsPlugin is such that it only looks for key of the destination and `All` key to handle the filtering logic. It will pass the event if it cannot find any of the above keys.

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [ ] I have checked for potential performance impacts and optimized if necessary.
- [ ] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

## How to test?

The sample destination in the sample app has the key `Amplitude`.
1. Set the integration options for some events before sending them as below
```kotlin
analytics.track(
    name = "Sample event",
    options = RudderOption(
        integrations = buildJsonObject {
            put("Amplitude", false)
        }
    )
)
```
2. Run the sample app.
3. Check for logs of those events. 
They should be passed or filtered out according to the integration options set.

## Maintainers Checklist

- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

### CODE REVIEW

File: android/src/test/kotlin/com/rudderstack/sdk/kotlin/android/plugins/devicemode/eventprocessing/IntegrationOptionsPluginTest.kt
Comment (1abhishekpandey) [inline android/src/test/kotlin/com/rudderstack/sdk/kotlin/android/plugins/devicemode/eventprocessing/IntegrationOptionsPluginTest.kt:1]: Great work with the test cases!

Let's also add few of the edge cases:

(Please rephrase/refactor them if needed)
```
@Test
fun `given an event with empty integrations, when plugin's intercept called with it, then it returns that event`() =
runTest(testDispatcher) {
    val event = TrackEvent(event = "event-name", properties = emptyJsonObject)
    event.integrations = emptyJsonObject

    val result = integrationOptionsPlugin.intercept(event)

    assertEquals(event, result)
}

@Test
fun `given an event with integration field set to string type, when plugin's intercept called with it, then it returns that event`() =
runTest(testDispatcher) {
    val event = TrackEvent(event = "event-name", properties = emptyJsonObject)
    event.integrations = buildJsonObject {
        put("MockDestination", "some string value")
    }

    val result = integrationOptionsPlugin.intercept(event)

    assertEquals(event, result)
}

@Test
fun `given an event with integration field set to complex type, when plugin's intercept called with it, then it returns that event`() =
runTest(testDispatcher) {
    val event = TrackEvent(event = "event-name", properties = emptyJsonObject)
    event.integrations = buildJsonObject {
        put("MockDestination", buildJsonObject {
            put("key", "value")
        })
    }

    val result = integrationOptionsPlugin.intercept(event)

    assertEquals(event, result)
}
```

Comment (1abhishekpandey) [reply android/src/test/kotlin/com/rudderstack/sdk/kotlin/android/plugins/devicemode/eventprocessing/IntegrationOptionsPluginTest.kt:1]: I can see your current logic doesn't handle above-edge cases!

File: android/src/main/kotlin/com/rudderstack/sdk/kotlin/android/plugins/devicemode/eventprocessing/IntegrationOptionsPlugin.kt
Comment (1abhishekpandey) [inline android/src/main/kotlin/com/rudderstack/sdk/kotlin/android/plugins/devicemode/eventprocessing/IntegrationOptionsPlugin.kt:1]: In the PR description, please add the implementation details and now since we are supporting the `Json` object for `integration` please highlight that!

Over here:
> 1. Set the integration options for some events before sending them.

Let's add a sample code snippet containing the `integration` object.

**Review by 1abhishekpandey — CHANGES_REQUESTED** [review]

**Review by ChryssaAliferi — COMMENTED** [review]
Align with Abhishek's comments on this one, so feel free to apply changes and request again for a review. 

🚀

File: android/src/main/kotlin/com/rudderstack/sdk/kotlin/android/plugins/devicemode/eventprocessing/IntegrationOptionsPlugin.kt [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    override lateinit var analytics: Analytics
+
+    override suspend fun intercept(event: Event): Event? {
+        val integrationOptions = event.integrations
+
+        integrationOptions.getBoolean(key)?.let { isExplicitlyEnabled ->
+            return when (isExplicitlyEnabled) {
+                true -> event
+                false -> {
+                    logDroppedEvent(event)
```
Comment (1abhishekpandey) [inline android/src/main/kotlin/com/rudderstack/sdk/kotlin/android/plugins/devicemode/eventprocessing/IntegrationOptionsPlugin.kt:45]: Nice!

**Review by 1abhishekpandey — APPROVED** [review]
Great work!
